internal/controllers/scheduling/controller.go

```diff
@@ -0,0 +1,195 @@
+}
+
+func (c *controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+	start := time.Now()
+	defer func() {
+		schedulingLatency.Observe(time.Since(start).Seconds())
+	}()
+
+	// Avoid conflict errors by waiting until we see the last dispatched synthesis (or timeout)
+	if c.lastApplied != nil {
+		ok, wait, err := c.lastApplied.HasBeenPatched(ctx, c.client, c.cacheGracePeriod)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("checking cache for previous op: %w", err)
+		}
+		if !ok {
+			logger.V(1).Info("waiting for cache to reflect previous operation")
+			return ctrl.Result{RequeueAfter: wait}, nil
+		}
+		c.lastApplied = nil
+	}
+
+	synths := &apiv1.SynthesizerList{}
+	err := c.client.List(ctx, synths)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing synthesizers: %w", err)
+	}
+	synthsByName, synthEpoch := indexSynthesizers(synths.Items)
+
+	comps := &apiv1.CompositionList{}
+	err = c.client.List(ctx, comps)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing compositions: %w", err)
+	}
+	nextSlot := c.getNextCooldownSlot(comps)
+
+	var inFlight int
+	var op *op
+	for _, comp := range comps.Items {
+		comp := comp
+		if comp.Synthesizing() {
+			inFlight++
+		}
+
+		synth, ok := synthsByName[comp.Spec.Synthesizer.Name]
+		if !ok {
+			continue
+		}
+
+		next := newOp(&synth, &comp)
+		if next != nil && (op == nil || op.Less(next)) {
+			op = next
+		}
+	}
+	freeSynthesisSlots.Set(float64(c.concurrencyLimit - inFlight))
+
+	if op == nil || inFlight >= c.concurrencyLimit {
+		return ctrl.Result{}, nil
+	}
+	if wait := time.Until(nextSlot); op.Reason.Deferred() && wait > 0 {
+		return ctrl.Result{RequeueAfter: wait}, nil
+	}
+	logger = logger.WithValues("compositionName", op.Composition.Name, "compositionNamespace", op.Composition.Namespace, "reason", op.Reason, "synthEpoch", synthEpoch)
+
+	// Maintain ordering across synth/composition informers by doing a 2PC on the composition
+	if op.Reason == synthesizerModifiedOp && setSynthEpochAnnotation(op.Composition, synthEpoch) {
+		if err := c.client.Update(ctx, op.Composition); err != nil {
+			return ctrl.Result{}, fmt.Errorf("updating synthesizer epoch: %w", err)
+		}
+		logger.V(1).Info("updated global synthesizer epoch")
+		return ctrl.Result{}, nil
+	}
+
+	if err := c.dispatchOp(ctx, op); err != nil {
+		if errors.IsInvalid(err) {
+			logger.V(0).Info("conflict while dispatching synthesis")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("dispatching synthesis operation: %w", err)
+	}
+
+	op.Dispatched = time.Now()
+	c.lastApplied = op
+	logger.V(0).Info("dispatched synthesis")
+
+	return ctrl.Result{}, nil
+}
+
+// getNextCooldownSlot returns the next time at which a deferred synthesis can be dispatched while honoring the configured cooldown period.
+func (c *controller) getNextCooldownSlot(comps *apiv1.CompositionList) time.Time {
+	var last time.Time
```

Comment:
Just FYI that I had to read this a few times to grok it, I was thining about `last` as the one with the oldest initialized. In my head `newest` would make more sense, but with this kind of things it gets subjective real fast...

Reply:
I think in this case you're just objectively correct 😆 . Fixed.